<?xml version='1.0'?>

<launch>

  <!-- agent properties -->
  <arg name='car_name' default='car_1' />
  <arg name='x_pos' default='0.0' />
  <arg name='y_pos' default='0.0' />
  <arg name='z_pos' default='0.05' />
  <arg name='run_gazebo' default='false' />
  <arg name='paint' default='Yellow' />
  <arg name='remote_control' default='false' />
  <arg name='keyboard_control' default='false' />
  <arg name='world_name' default='bahrain_track' />
  <arg name='mapping_tf' default='false' />

  <!-- launch map to odom tf -->
  <group unless='$(arg mapping_tf)'>
    <node name='map_$(arg car_name)_odom_link' pkg='tf' type='static_transform_publisher' args='$(arg x_pos) $(arg y_pos) 0.0
                                  0.0 0.0 0.0
                                  map
                                  $(arg car_name)_odom
                                  100' />
  </group>

  <!-- run gazebo by default only once -->
  <group if='$(arg run_gazebo)'>

    <!-- global parameters -->
    <param name='use_sim_time' value='true' />

    <!-- simulator arguments -->
    <arg name='gui' default='true' />
    <arg name='run_camera' default='false' />

    <!-- launch gazebo with track -->
    <include file='$(find gazebo_ros)/launch/empty_world.launch'>
      <arg name='world_name' value='$(find f1tenth-sim)/world/$(arg world_name).world' />
      <arg name='gui' value='$(arg gui)' />
    </include>
  </group>

  <!-- check if remote control is required -->
  <group if='$(arg remote_control)'>
    <node name='joy_plugin' pkg='joy' type='joy_node'>
      <param name='autorepeat_rate' value='20.0' />
    </node>
    <node name='remote_control' pkg='f1tenth-sim' type='remote_control.py' args='$(arg car_name)' />
  </group>

  <!-- check if keyboard control is required -->
  <group if='$(arg keyboard_control)'>
    <node name='keyboard_plugin' pkg='f1tenth-sim' type='keyboard_teleop.py' args='$(arg car_name)' />
  </group>

  <!-- reset racecar to pit stop -->
  <node name='$(arg car_name)_pit_stop_reset' pkg='f1tenth-sim' type='set_racecar_state.py' args='$(arg car_name)' />

  <!-- spawn car using the set global parameters -->
  <include file='$(find f1tenth-sim)/launch/car_spawner.launch'>
    <arg name='car_name' value='$(arg car_name)' />
    <arg name='x_pos' value='$(arg x_pos)' />
    <arg name='y_pos' value='$(arg y_pos)' />
    <arg name='z_pos' value='$(arg z_pos)' />
    <arg name='paint' value='$(arg paint)' />
  </include>

</launch>
